server/src/main/java/com/paypal/selion/proxy/SeLionSauceProxy.java

```
@@ -22,72 +22,248 @@
             } catch (InterruptedException e) {
                 LOGGER.log(Level.SEVERE, e.getMessage(), e);
             }
-            if (sauceApi.getNumberOfTCRunning() <= getMaxNumberOfConcurrentTestSessions()) {
-                String username = (String) requestedCapability.get("sauceUserName");
-                String accessKey = (String) requestedCapability.get("sauceApiKey");
-                String tunnelId = (String) requestedCapability.get("parent-tunnel");
-                if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(accessKey)) {
-                    requestedCapability.put("username", username);
-                    requestedCapability.put("accessKey", accessKey);
-                }
-                if (StringUtils.isNotEmpty(tunnelId)) {
-                    requestedCapability.put("parent-tunnel", tunnelId);
-                }
-                session = super.getNewSession(requestedCapability);
+
+            // check for backwards compatible capability names that might come across the wire
+            // TODO remove backwards compatible support in a future version.
+            processDeprecatedCapabilities(requestedCapabilities);
+
+            // default the credentials, if required
+            processCredentails(requestedCapabilities);
+
+            // setup the tunnelIdentifier, if required
+            processTunnelIdentifierCapability(requestedCapabilities);
+
+            // setup the parent-tunnel, if required
+            processParentTunnelCapability(requestedCapabilities);
+
+            // override any idleTimeout capability that may have been specified. Sauce and this proxy need to have
+            // a similar idle timeout value.
+            requestedCapabilities.put(SauceLabsCapability.IDLE_TIMEOUT_CAPABILITY, getTimeOut() / 1000);
+
+            // if call to geNumberOfTCRunning returns -1, sauce REST API calls are failing.
+            // so we kick the request back to the hub.
```

... thinking more about this, there might be a better way to fail faster on the client-side. for example; if `numberOfTCRunning == -1`, this means we couldn't talk to sauce after the `retry` limit was exhausted.. So, we can just mark it as `down=true` (which ideally will trigger the line 229) for the whole proxy and then let the polling thread figure out when it is back up... I'll see what I can do.